### PR TITLE
WIP:  File Watcher gateway as core gateway, notification filters to storage grid gateway, k8 configmap as trigger artifact location.

### PR DESCRIPTION
**Pull request by VaibhavPage**
- [x] Correct ordering of gateway container startup. Gateway transformer > processor
- [x] Add K8 configmaps as artifact location for triggers
- [x] Add filters to storage grid notifications
- [x] Add file watcher as core gateway

#106 #104 #98

### CODE REVIEW

**Comment by magaldima** [comment]
please check the unit tests, I'm seeing 
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x142c643]
```

for the `github.com/argoproj/argo-events/controllers/gateway` package `TestProcessNextItem` test.

File: controllers/gateway/transform/transformer.go [hunk 1]
```diff
@@ -286,3 +286,9 @@ func (toc *tOperationCtx) TransformRequest(w http.ResponseWriter, r *http.Reques
 	}
 	common.SendSuccessResponse(w)
 }
+
+// ReadinessProbe is probe to check whether server is running or not
+func (toc *tOperationCtx) ReadinessProbe(w http.ResponseWriter, r *http.Request) {
+	toc.log.Info().Msg("transformer is up and running")
```
Comment (magaldima) [inline controllers/gateway/transform/transformer.go]: do we need to log a message? we can remove to clean up some logging

File: gateways/core/webhook/webhook.go [hunk 2]
```diff
@@ -200,13 +200,17 @@ func (wce *webhookConfigExecutor) StopConfig(config *gateways.ConfigContext) err
 }
 
 func main() {
-	_, err := gatewayConfig.WatchGatewayEvents(context.Background())
+	err := gatewayConfig.TransformerReadinessProbe()
```
Comment (magaldima) [inline gateways/core/webhook/webhook.go]: is there any way to reduce some of this boilerplate code (possibly centralize to a util method).. I'm seeing this code and corresponding changes for all gateway types...

File: gateways/custom/storage-grid/main.go [hunk 3]
```diff
@@ -61,26 +66,70 @@ type storageGridEventConfig struct {
...
 	Events []string
-	Filter StorageGridFilter
+	Filter *Filter
 	// srv holds reference to http server
 	srv *http.Server
 	mux *http.ServeMux
 }
 
+// Filter represents filters to apply to bucket nofifications for specifying constraints on objects
+type Filter struct {
```
Comment (magaldima) [inline gateways/custom/storage-grid/main.go]: been thinking more about this kind of stuff.. the user needs to know the structure of all these types when he/she builds the appropriate configmap for a gateway.. don't you think this belongs in the API package somewhere? But at the very least, how do we ensure this stuff is transparent, and validated against whenever a user creates a config?

Comment (VaibhavPage) [reply gateways/custom/storage-grid/main.go]: I agree that this needs to be transparent. We can add `Validate` method to `ConfigExecutor` interface to make sure that configs are validated.

Comment (VaibhavPage) [reply gateways/custom/storage-grid/main.go]: Also a separate `types.go` file containing structs might help with readability

**Pull request merged**